src/frontend/Typechecker.ml

```diff
@@ -1001,40 +1002,178 @@ let warn_self_assignment loc lhs rhs =
+              | Unequal_lengths -> err lhs Equals rhs
+              | Ok l -> List.unzip l in
+            if
+              List.exists ~f:(function NoPromotion -> false | _ -> true) proms
+            then (TuplePromotion proms, TupleAD ad_levels)
+            else (NoPromotion, TupleAD ad_levels)
+        | LTuplePack _, rhs -> err lhs Equals rhs in
+      let prom, ad_level = typechk lhs rhs.emeta.type_ in
+      let rhs =
+        (* Hack: need RHS to properly get promoted to var if needed *)
+        {rhs with emeta= {rhs.emeta with ad_level}} in
+      Promotion.promote rhs prom
   | OperatorAssign op -> (
-      let args = List.map ~f:arg_type [Ast.expr_of_lvalue lhs; rhs] in
+      let args =
+        [(UnsizedType.AutoDiffable, type_of_lvalue lhs); arg_type rhs] in
       let return_type = assignmentoperator_stan_math_return_type op args in
-      match return_type with Some Void -> rhs | _ -> err op |> error )
+      match return_type with
+      | Some Void -> rhs
+      | _ -> err lhs op rhs.emeta.type_ )
+
+let overlapping_lvalues lvals =
+  (* Prevent assigning to multiple indices in the same object at once.
+     In principal it is safe to assign to disjoint indices, but statically
+     checking that is impossible. *)
+  let rec compare_no_indexing lv1 lv2 =
+    match (lv1.lval, lv2.lval) with
+    | LIndexed (l, _), _ -> compare_no_indexing l lv2
+    | _, LIndexed (l, _) -> compare_no_indexing lv1 l
+    | LVariable id1, LVariable id2 -> String.compare id1.name id2.name
+    | LTupleProjection (lv1, idx1), LTupleProjection (lv2, idx2)
+      when idx1 = idx2 ->
+        compare_no_indexing lv1 lv2
+    | _, _ ->
+        (* remaining cases are not equal, we don't care *)
+        Ast.compare_untyped_lval lv1 lv2 in
+  List.find_all_dups lvals ~compare:compare_no_indexing
+
+let lvalues_written_to lv =
+  let rec add_tuple_idxs lv : typed_lval list =
+    (* If we're assigning an entire tuple, we also need to prevent
+       assigning to any slot in this statement. *)
+    let type_, _ = UnsizedType.unwind_array_type lv.lmeta.type_ in
+    match (lv.lval, type_) with
+    | _, UTuple ts ->
+        List.concat_mapi ts ~f:(fun i ty ->
+            add_tuple_idxs
+              { lval= LTupleProjection (lv, i + 1)
+              ; lmeta= {lv.lmeta with type_= ty} } )
+    | _ -> [lv] in
+  let rec flatten_lvalue_pack lv =
+    match lv with
+    | LTuplePack (lvs, _) -> List.concat_map ~f:flatten_lvalue_pack lvs
+    | LValue lv -> [lv] in
+  flatten_lvalue_pack lv
+  |> List.concat_map ~f:add_tuple_idxs
+  |> List.map ~f:Ast.untyped_lvalue_of_typed_lvalue
+
+let variables_accessed_in lv =
+  (* We only care about values being read inside the lvalue here,
+     which means only the expressions inside of LIndexed *)
+  let rec extract_indices lv =
+    match lv.lval with
+    | LVariable _ -> []
+    | LTupleProjection (lv, _) -> extract_indices lv
+    | LIndexed (lv, es) -> extract_indices lv @ es in
```

Comment:
If you're going to call `List.dedup_and_sort` at the end then you could use a `Set` instead of `List` in the first place.
```suggestion
  let rec extract_indices lv =
    match lv.lval with
    | LVariable _ -> String.Set.empty
    | LTupleProjection (lv, _) -> extract_indices lv
    | LIndexed (lv, es) ->
        List.concat_map ~f:exprs_in_index es
        |> List.concat_map ~f:extract_ids
        |> List.map ~f:(fun {name; _} -> name)
        |> String.Set.of_list
        |> Set.union (extract_indices lv) in
```